fuzzers/main.py

```diff
@@ -0,0 +1,337 @@
+import datetime
+import google.cloud.storage
+from http.server import BaseHTTPRequestHandler, HTTPServer
+import inotify.adapters
+import os
+import pathlib
+import prometheus_client
+import random
+import signal
+import subprocess
+import sys
+import threading
+import time
+import toml
+import uuid
+
+from lib import config
+from workers import utils
+
+WORKDIR = utils.WORKDIR
+REPO_URL = utils.REPO_URL
+#WORKDIR = pathlib.Path('/tmp')
+#REPO_URL = 'https://github.com/Ekleog/nearcore'
+def connect_to_gcs():
+    subprocess.run(['gcloud', 'auth', 'activate-service-account', '--key-file', GCS_CREDENTIALS_FILE], check=True)
+    return google.cloud.storage.Client.from_service_account_json(GCS_CREDENTIALS_FILE)
+    #return google.cloud.storage.Client(project = 'near-nayduck')
+
+NUM_FUZZERS = os.cpu_count()
+#NUM_FUZZERS = 1
+
+CMD_PORT = 7055
+METRICS_PORT = 5507
+
+REPO_DIR_PARENT = WORKDIR
+REPO_DIR_NAME = 'fuzzed-nearcore'
+REPO_DIR = REPO_DIR_PARENT / REPO_DIR_NAME
+
+GCS_CREDENTIALS_FILE = config.CONFIG_DIR / 'credentials.json'
+GCS_BUCKET = 'fuzzer'
+
+FUZZ_TIME = prometheus_client.Counter('fuzz_seconds', 'Time spent fuzzing', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CRASHES = prometheus_client.Counter('fuzz_crashes', 'Number of times the fuzzer process crashed (not unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_ARTIFACTS_FOUND = prometheus_client.Counter('fuzz_artifacts_found', 'Number of artifacts found (should be number of unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_UPLOADED = prometheus_client.Counter('fuzz_corpus_uploaded', 'Number of elements uploaded to GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_DELETED = prometheus_client.Counter('fuzz_corpus_deleted', 'Number of elements deleted from GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+
+def update_repo(branch):
+    if not REPO_DIR.exists():
+        print(f"Doing initial clone of repository {REPO_DIR}")
+        if 0 != subprocess.run(['git', 'clone', REPO_URL, REPO_DIR_NAME], cwd=REPO_DIR_PARENT).returncode:
+            raise RuntimeError(f"Failed to clone from repo {REPO_URL}")
```

Comment:
By the way, you might look into using workers.utils.Runner class.  It is a bit awkward perhaps but if it does what you need than it’ll automatically print commands being executed to standard error.

Reply:
In this case it's convenient for me to keep the stdout/stderr piped to the top-level stdout/err, because this is a process-wide information. As for the places where we run rsync and/or the fuzzer, it'd probably work well if there were a way to tell Runner to send both stdout and stderr to a single file, to make things easier on the reader with a single log file to get from GCS, though I'm not sure it'd bring much compared to just using subprocess.*?

Reply:
One thing Runner gives you though is that it prints the command that it is executing. This may be helpful when debugging. If you instantiate Runner without outdir parameter it will send all the logs to stdout/stderr. I’m not saying that using it is definitely a good idea, but it may be something worth looking a few minutes at.

Reply:
Hmm but the documentation of `__init__` says “Otherwise, creates two temporary files and redirects output there.” and the code appears to match this documentation?

Reply:
Ah, you’re right. Not sure how I missed that. I wasn’t completely crazy though, the class does print executed commands to stderr.